Add required phrase rules automatically

AyanSinhaMahapatra opened the pull request:
Reference: #2637 #2878 

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Add script and functions for adding required phrases

Add a script which can add required phrases in already existing rules
automatically from license field names, and can do this one license
expression at a time.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for `apache-2.0` rules

These were added automatically by the script added at
etc/scripts/licenses/add_required_phrases.py

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for `apache-1.1` rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectaion

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into required-phrase-rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into required-phrase-rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Move required-phrases script to licensedcode

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add a new console script to add required phrases

* moves the script to licensedcode
* adds a new console script `add-required-phrases`
* move the functions and tests to seperate files

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update apache-2.0 rules with required phrases

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into required-phrase-rules

AyanSinhaMahapatra committed:
Add required phrases from other rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -2844,83 +2896,39 @@ def find_rule_base_location(name_prefix, rules_directory=rules_data_dir):
+        if rule.license_expression:
+            rules_by_expression[rule.license_expression].append(rule)
+
+    return OrderedDict(sorted(rules_by_expression.items()))
```
You can use a plain dict to the same effect.

AyanSinhaMahapatra replied:
Thanks, resolved.

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1425,33 @@ class BasicRule:
+        default=False,
+        repr=False,
+        metadata=dict(
+            help='True if this is rule text is a required phrase '
```
```suggestion
            help='True if this is rule text is a required phrase. '
```

AyanSinhaMahapatra replied:
Added, thanks!

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1425,33 @@ class BasicRule:
+            'A required phrase is often a part of another larger rule text '
+            'but is an essential section of the rule text which must be '
+            'present in the case of partial matches, otherwise the match '
+            'will be a false positive and misleading. ')
```
```suggestion
            'will be a false positive and misleading.')
```

AyanSinhaMahapatra replied:
Added, thanks!

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1425,33 @@ class BasicRule:
+        metadata=dict(
+            help='True if this rule needs to be skipped while creating '
+            'required phrase rules. Required phrase rules are created out '
+            'of other rule texts which have marked required phrases. ')
```
```suggestion
            'of other rule texts which have marked required phrases.')
```

AyanSinhaMahapatra replied:
Added, thanks!

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1796,22 +1813,27 @@ def has_unknown(self):
             self.is_license_tag,
             self.is_license_intro,
             self.is_license_clue,
+            self.is_required_phrase,
```
IMHO this is not a license flag

AyanSinhaMahapatra replied:
Removed, thanks!

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1796,22 +1813,27 @@ def has_unknown(self):
-        is_false_positive = self.is_false_positive
-
-        license_flags = (
+    @property
```
If is_required_phrase is not a flag, then the update to validate may no longer be needed?

AyanSinhaMahapatra replied:
Used elsewhere to make things compact, keeping this.

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1425,33 @@ class BasicRule:
+            'will be a false positive and misleading. ')
+    )
+
+    skip_creating_required_phrases = attr.ib(
```
What about instead: skip_collecting_required_phrases

AyanSinhaMahapatra replied:
This is better, added this, thanks!

pombredanne commented on src/licensedcode/tokenize.py:
```diff
@@ -113,6 +115,117 @@ def key_phrase_tokenizer(text, stopwords=STOPWORDS):
+    return combined_text
+
+
+def add_key_phrase_markers(text, key_phrase_span):
```
Use only one of key_phrases or required_phrases

AyanSinhaMahapatra replied:
Renaming `key phrase` -> `required phrase` everywhere as this is more verbose.

pombredanne reviewed (commented):
Here are some comments for your consideration.

AyanSinhaMahapatra committed:
Update and improve collecting and marking required phrases

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add changes on marking required phrases automatically

This is done only on `mit` rules, by running:
add-required-phrases -ovr --license-expression "mit"

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures and address feedback

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/licensedcode/tokenize.py:
```diff
@@ -103,17 +105,133 @@ def key_phrase_tokenizer(text, stopwords=STOPWORDS):
+    tokens_tuples_with_markers = []
+    token_index = 0
+
+    for token_tuple in tokens_tuples_without_markers:
```
What do you gain from an intermediate list?
What about this?
```suggestion
    for token_tuple in matched_query_text_tokenizer(text):
```

pombredanne commented on src/licensedcode/tokenize.py:
```diff
@@ -103,17 +105,133 @@ def key_phrase_tokenizer(text, stopwords=STOPWORDS):
+        _value, token = token_tuple
+        combined_text += token
+        
+    return combined_text
```
What about this, without intermediate accumulation of strings?
```suggestion
    return ''.join(token for _, token in token_tuples)
```

pombredanne commented on src/licensedcode/tokenize.py:
```diff
@@ -103,17 +105,133 @@ def key_phrase_tokenizer(text, stopwords=STOPWORDS):
+        return new_required_phrase_spans
+    
+    for new_span in new_required_phrase_spans:
+        if not any(
```
This is a double negative... what about:
```suggestion
        if any(old_span.overlap(new_span) for old_span in old_required_phrase_spans:
            continue
```
or 
```suggestion
        if not any(old_span.overlap(new_span) for old_span in old_required_phrase_spans:
            yield ...
```

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1796,24 +1814,37 @@ def has_unknown(self):
+        }
+
+    @property
+    def license_flag_values(self):
```
since this is used only twice, is this needed?

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1866,6 +1897,12 @@ def validate(self, licensing=None, thorough=False):
             if not check_is_list_of_strings(self.referenced_filenames):
                 yield 'referenced_filenames must be a list of strings'
 
+            if (self.is_license_clue or self.is_license_intro) and self.is_required_phrase:
```
When is this an issue?

AyanSinhaMahapatra replied:
This is to prevent marking of license clues or license intro rules as required phrases as by definition these two types of rules are either non-specific intros or weak clues as rules, which cannot be considered required phrases for any specific license expressions.

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1796,24 +1814,37 @@ def has_unknown(self):
-        has_license_flags = any(license_flags)
-        has_many_license_flags = len([l for l in license_flags if l]) != 1
+        has_license_flags = any(self.license_flag_values)
+        has_many_license_flags = len([l for l in self.license_flag_values if l]) > 1
```
This is minor, but what about this if you remove the self.license_flag_values method above?
```suggestion
        license_flags_count = sum(self.license_flag.values()
        has_many_license_flags = license_flags_count > 1
        has_license_flags = boolean(license_flags_count)

```

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -2260,6 +2291,8 @@ def dump(self, rules_data_dir):
         rule_file = self.rule_file(rules_data_dir=rules_data_dir)
 
         metadata = self.to_dict()
+        if kwargs:
```
Can you explain in a comment this is used only in debug and possibly prefix kwargs with an underscore?

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -2291,6 +2324,9 @@ def load(self, rule_file, with_checks=True):
             raise e
 
         known_attributes = set(attr.fields_dict(self.__class__))
+        # This is an attirbute used to debug marking required phrases, and is not needed
```
```suggestion
        # This is an attribute used to debug marking required phrases, and is not needed
```

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1424,35 @@ class BasicRule:
+            'but is an essential section of the rule text which must be '
+            'present in the case of partial matches, otherwise the match '
+            'will be a false positive and misleading.')
+    )
```
I wonder if this would not be best computed an index time ... this is True if a rule text is entirely a required phrase, this would mean one less stored flag.

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1428,13 +1424,35 @@ class BasicRule:
+        default=False,
+        repr=False,
+        metadata=dict(
+            help='True if this is rule text is a required phrase. '
```
I am still unsure about this field... is this when the whole rule text is a required phrase? Because looking at this again now, I wonder if this would not be simpler to generate the flag when we index a rule based on this fact.

AyanSinhaMahapatra replied:
There are a couple reasons why we are doing it this way, explicitly with a flag, instead of computing these on the fly:

1. We are not generating these required phrase rules while loading the index from the rules, this is instead an one time operation and a thing that only runs on tests or new rule additions, where we go thorough the rules collect required phrases and mark them in other rules (while also saving them, by creating a new file, or already marking an existing rule with the flag). If we start generating this on the fly while creating the index we are making the index creation step more expensive (we now only collect required phrases from a single rule, but if we do this we have to also add them as rules to the main index) and it will take even more time.
2. Today the process to create required phrase rules is not fully automated at first, we need to review all the newly created required phrase rules and then mark some origin rules with `skip_collecting_required_phrases: yes` so that some rule-specific required phrases are not created as required phrases. This process will become much more abstract and non-intuitive, and would be much more difficult to debug.
3. The license detection process update will also use these new required phrase rules. In cases where we don't have good matches, we will try to do a license detection just filtering required phrase rules from all the matches, and this would involve looking at all matches which has `is_required_phrase` as True, therefore making this flag necessary.

pombredanne replied:
OK.Thanks!

pombredanne reviewed (commented):
Some minor nits and a question on the is_required_phrase flag for your consideration!

AyanSinhaMahapatra committed:
Fix required phrase span bugs

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne reviewed (approved):
Looking fine!

AyanSinhaMahapatra commented:
Closing this as we now have https://github.com/aboutcode-org/scancode-toolkit/pull/3924
